test/e2e/e2e_local_source_upload_test.go

```diff
@@ -0,0 +1,104 @@
+	"github.com/shipwright-io/build/test/utils"
+)
+
+var _ = Describe("For a Kubernetes cluster with Tekton and build installed", func() {
+	var (
+		testID   string
+		build    *buildv1alpha1.Build
+		buildRun *buildv1alpha1.BuildRun
+	)
+
+	AfterEach(func() {
+		if buildRun != nil {
+			testBuild.DeleteBR(buildRun.Name)
+			buildRun = nil
+		}
+
+		if build != nil {
+			testBuild.DeleteBuild(build.Name)
+			build = nil
+		}
+	})
+
+	Context("when LocalCopy BuildSource is defined", func() {
+		BeforeEach(func() {
+			testID = generateTestID("local-copy")
+			build = createBuild(
+				testBuild,
+				testID,
+				"test/data/build_buildah_cr_local_source_upload.yaml",
+			)
+		})
+
+		It("should generate LocalCopy TaskRun, using the waiter", func() {
+			var err error
+			buildRun, err = buildRunTestData(
+				testBuild.Namespace,
+				testID,
+				"test/data/buildrun_buildah_cr_local_source_upload.yaml",
+			)
+			Expect(err).ToNot(HaveOccurred(), "Error retrieving buildrun test data")
+
+			validateWaiterBuildRun(testBuild, buildRun)
+		})
+	})
+})
+
+func getBuildRunStatus(name types.NamespacedName) *corev1.ConditionStatus {
+	testBuildRun, err := testBuild.LookupBuildRun(name)
+	Expect(err).ToNot(HaveOccurred(), "Error retrieving the BuildRun")
+
+	if len(testBuildRun.Status.Conditions) == 0 {
+		return nil
+	}
+	return &testBuildRun.Status.GetCondition(buildv1alpha1.Succeeded).Status
+}
+
+// validateWaiterBuildRun assert the BuildRun informed will fail, since Waiter's timeout is reached
+// and it causes the actual build process to fail as well.
+func validateWaiterBuildRun(testBuild *utils.TestBuild, testBuildRun *buildv1alpha1.BuildRun) {
+	// making sure the BuildRun has been created on the cluster
+	if _, err := testBuild.GetBR(testBuildRun.Name); err != nil {
+		Expect(testBuild.CreateBR(testBuildRun)).
+			ToNot(HaveOccurred(), "Failed to create BuildRun")
+	}
+
+	falseCondition := corev1.ConditionFalse
+	unknownCondition := corev1.ConditionUnknown
+
+	buildRunName := types.NamespacedName{Namespace: testBuild.Namespace, Name: testBuildRun.Name}
+
+	// making sure the taskrun is schedule and becomes a pod, since the build-controller will transit
+	// the object status from empty to unknown, when the actual build starts
+	Eventually(func() *corev1.ConditionStatus {
+		status := getBuildRunStatus(buildRunName)
+		Logf("BuildRun '%s' status '%v'...", buildRunName, status)
+		return status
+	}, time.Duration(60*time.Second), 3*time.Second).
+		Should(Equal(&unknownCondition), "BuildRun should reach unknown condition")
+
+	// asserting the waiter step will end up in timeout, in other words, it ends as failed
+	Eventually(func() *corev1.ConditionStatus {
+		status := getBuildRunStatus(buildRunName)
+		Expect(status).ToNot(BeNil())
+		Logf("BuildRun '%s' status '%s'...", buildRunName, *status)
+		return status
+	}, time.Duration(90*time.Second), 10*time.Second).
+		Should(Equal(&falseCondition), "BuildRun should end up in timeout")
```

Comment:
I suggest to wait for the Reason value (Running in the first block, Failed in the second one) and assert the reason. Basically it will be unknown already when the BuildRun is Pending and it can theoretically stay there for a while (waiting for the pod to be scheduled and its images pulled).

Reply:
Please consider latest changes, I'm using your suggestion and checking the actual "Reason", inside the Condition slice.

I wonder if we should also extend the 90s timeout to count for the image pulling time as well.

Reply:
Right, especially with a lower bandwidth, it might not be enough. For this purpose, there is also the timeout multiplyer that you could add here as well. Like here: https://github.com/shipwright-io/build/blob/973ef833e68edb807658df3d1a20a55d75ffebf2/test/e2e/validators_test.go#L140

Reply:
Thanks for the snippet, it's in place now.